File: archive/v1/03_search_icon.py
import os

import onnxruntime as ort
import numpy as np
from transformers import AutoTokenizer
from scipy.spatial.distance import cosine
import json
import dotenv

dotenv.load_dotenv()

# Load the ONNX model
ort_session = ort.InferenceSession("../../data/multilingual_e5_quantized.onnx")

# Load the tokenizer with the corrected model name
tokenizer = AutoTokenizer.from_pretrained(os.environ.get("EMBEDDING_MODEL"))

# Load precomputed icon embeddings and metadata
icon_embeddings = np.load("../../data/icon_embeddings.npy")  # Shape: (num_icons, embedding_dim)
with open("../../data/icons_index.json", "r", encoding="utf-8") as f:
    icon_metadata = json.load(f)  # List of dictionaries with "id" and "tags"

# Function to generate an embedding from text
def get_embedding(text):
    # Tokenize the input text
    inputs = tokenizer(text, return_tensors="np", padding="max_length", max_length=128, truncation=True)
    input_ids = inputs["input_ids"].astype(np.int64)  # Shape: (1, 128)
    attention_mask = inputs["attention_mask"].astype(np.int64)  # Shape: (1, 128)

    # Run inference with the ONNX model
    ort_inputs = {"input_ids": input_ids, "attention_mask": attention_mask}
    embedding = ort_session.run(None, ort_inputs)[0]  # Output shape: (1, embedding_dim)
    return embedding.flatten()  # Flatten to 1D array: (embedding_dim,)

# Function to search icons based on a query
def search_icons(query, top_k=5):
    # Generate embedding for the user query
    query_embedding = get_embedding(query)

    # Compute cosine similarity between query and icon embeddings
    similarities = []
    for icon_emb in icon_embeddings:
        sim = 1 - cosine(query_embedding, icon_emb)  # 1 - cosine distance = similarity
        similarities.append(sim)

    # Get the top-k most similar icons
    top_indices = np.argsort(similarities)[::-1][:top_k]  # Sort descending, take top_k
    results = [icon_metadata[i] for i in top_indices]
    return results

# Example usage
user_query = input("Enter some search term: ")
results = search_icons(user_query)
for icon in results:
    print(f"Icon: {icon['id']}, Data: {icon}")

File: archive/v1/03_search_icon.swift
import Foundation
import ONNXRuntime

// Helper function to compute cosine similarity
func cosineSimilarity(_ a: [Float], _ b: [Float]) -> Float {
    let dotProduct = zip(a, b).map(*).reduce(0, +)
    let normA = sqrt(a.map { $0 * $0 }.reduce(0, +))
    let normB = sqrt(b.map { $0 * $0 }.reduce(0, +))
    return dotProduct / (normA * normB)
}

// Load precomputed icon embeddings and metadata from JSON
func loadData() -> ([[Float]], [[String: String]]) {
    guard let embeddingsURL = Bundle.main.url(forResource: "icon_embeddings", withExtension: "json"),
          let metadataURL = Bundle.main.url(forResource: "icons_index", withExtension: "json"),
          let embeddingsData = try? Data(contentsOf: embeddingsURL),
          let metadataData = try? Data(contentsOf: metadataURL),
          let embeddings = try? JSONDecoder().decode([[Float]].self, from: embeddingsData),
          let metadata = try? JSONDecoder().decode([[String: String]].self, from: metadataData) else {
        fatalError("Failed to load data")
    }
    return (embeddings, metadata)
}

// Simple pre-tokenized input (for demo; replace with actual tokenization if needed)
func getTokenizedInput(query: String) -> (inputIds: [Int64], attentionMask: [Int64]) {
    // Placeholder: In practice, use a tokenizer or precomputed tokens
    // Here, we simulate tokenized input with dummy values
    let inputIds = Array(repeating: Int64(query.count % 1000), count: 128) // Simplified
    let attentionMask = Array(repeating: Int64(1), count: 128)
    return (inputIds, attentionMask)
}

// Load and run the ONNX model
func getEmbedding(inputIds: [Int64], attentionMask: [Int64]) -> [Float] {
    guard let modelPath = Bundle.main.path(forResource: "embedder", ofType: "onnx") else {
        fatalError("Model file not found")
    }

    let session = try! ORTSession(modelPath: modelPath)
    let inputIdsTensor = try! ORTTensor(int64Values: inputIds, dimensions: [1, 128])
    let attentionMaskTensor = try! ORTTensor(int64Values: attentionMask, dimensions: [1, 128])

    let inputs: [String: ORTTensor] = [
        "input_ids": inputIdsTensor,
        "attention_mask": attentionMaskTensor
    ]

    let outputs = try! session.run(withInputs: inputs)
    return outputs["output"] as! [Float]  // Adjust "output" to match your model’s output name
}

// Search function
func searchIcons(query: String, topK: Int = 5) -> [[String: String]] {
    let (iconEmbeddings, iconMetadata) = loadData()
    let (inputIds, attentionMask) = getTokenizedInput(query: query)
    let queryEmbedding = getEmbedding(inputIds: inputIds, attentionMask: attentionMask)

    var similarities: [(Int, Float)] = []
    for (index, iconEmb) in iconEmbeddings.enumerated() {
        let sim = cosineSimilarity(queryEmbedding, iconEmb)
        similarities.append((index, sim))
    }

    let topIndices = similarities
        .sorted { $0.1 > $1.1 }
        .prefix(topK)
        .map { $0.0 }

    return topIndices.map { iconMetadata[$0] }
}

// Example usage
let query = "star"
let results = searchIcons(query: query)
for icon in results {
    print("Icon ID: \(icon["id"] ?? ""), Tags: \(icon["tags"] ?? "")")
}